### PR TITLE
Improve project documentation

### PR DESCRIPTION
**Pull request by Sharcoux**
I improved the error messages to help developers know exactly what was the issue.

I used JSDoc and ts-check annotation to enable IDE to extract types and do type checking in the files. Developers using typescript or IDE will see description for the methods and their parameters, and get automatic type checking.

I fixed an issue in import method. You probably messed up between using `value === undefined` and `typeof value === 'undefined'`

### CODE REVIEW

**Comment by Sharcoux** [comment]
image

File: assets/scripts/whichx.js [hunk 1]
```diff
@@ -158,18 +201,29 @@ function WhichX(config) {
...
         if (typeof description === "string") {
             // Remove special characters.
-            description = description.replace(/[^a-zA-Z ]/g, "");
+            description = description
+                .normalize("NFD")
+                .toLowerCase()
+                .replace(/[\u0300-\u036f]/g, "")
+                .replace(/[^a-zA-Z ]/g, "");
             // Lower case.
             description = description.toLowerCase();
```
Comment (rudikershaw) [inline assets/scripts/whichx.js]: This `toLowerCase()` now appears to be a duplicate.

File: assets/scripts/whichx.js [hunk 2]
```diff
@@ -158,18 +201,29 @@ function WhichX(config) {
...
+     * @param {string} description The description to analyse
+     * @returns {string[]} The list of normalized words contained in the description
+     */
     function processToArray(description) {
         var i = 0;
         if (typeof description === "string") {
             // Remove special characters.
-            description = description.replace(/[^a-zA-Z ]/g, "");
+            description = description
+                .normalize("NFD")
```
Comment (rudikershaw) [inline assets/scripts/whichx.js]: `.normalize()` is not available in Internet Explorer. Before this addition WhichX supports IE9 and above. Can you find an alternative? I appreciate this may be a big ask.

File: assets/scripts/whichx.js [hunk 3]
```diff
@@ -115,27 +147,38 @@ function WhichX(config) {
...
-        if (newTotal === "undefined" || newTotal.tcount === "undefined" || newTotal.wordTotal === "undefined") {
+        if (newTotal === undefined || newTotal.tcount === undefined || newTotal.wordTotal === undefined) {
             throw new Error("Import invalid. This doesn't look like it was exported from a prior model.");
         }
         typesMap = importedTypesMap;
     };
 
-    // Loop through words and work out probability of type given each word.
-    // Multiply each word's probability by total probability to determine type probability.
+    /**
```
Comment (rudikershaw) [inline assets/scripts/whichx.js:175]: `getTypeChance()` is not available to call on the `WhichX` object, so the JSDoc here may not be all the helpful. Can we remove the docs to keep file size down?

File: assets/scripts/whichx.js [hunk 4]
```diff
@@ -158,18 +201,29 @@ function WhichX(config) {
         return typeChance * (type.tcount / total.tcount);
     }
 
-    // A non-zero prior estimate to prevent 0 based probability calculations.
+    /**
```
Comment (rudikershaw) [inline assets/scripts/whichx.js:204]: This function is not available to call on the `WhichX` object, so the JSDoc here may not be all the helpful. Can we remove the docs to keep file size down?

File: assets/scripts/whichx.js [hunk 5]
```diff
@@ -158,18 +201,29 @@ function WhichX(config) {
...
+     * A non-zero prior estimate to prevent 0 based probability calculations.
+     * @returns {number} The non-zero probablity
+     */
     function mEstimate() {
         var total = typesMap.total;
         return 1 / (total.wordTotal * 100);
     }
 
-    // Process the description into an array of pertinent standardized lower case words.
```
Comment (rudikershaw) [inline assets/scripts/whichx.js:167]: This function is not available to call on the `WhichX` object, so the JSDoc here may not be all the helpful. Can we remove the docs to keep file size down?

**Review by rudikershaw — CHANGES_REQUESTED** [review]
Thank you so much for your contribution. There are some great improvements in here. In addition to the other comments, could you also;

* Add tests to demonstrate & test the Unicode normalization and diacritic removal.
* Add or alter tests to test the new and more specific errors for labels and descriptions.

File: assets/scripts/whichx.js (hunk 1 above)
Comment (Sharcoux) [reply assets/scripts/whichx.js]: Will fix

File: assets/scripts/whichx.js (hunk 2 above)
Comment (Sharcoux) [reply assets/scripts/whichx.js]: The current solution is to rely on [unorm](https://github.com/walling/unorm/). Is it ok? Other solution would require to maintain an exhaustive list of letter characters from all languages... We could also decide to keep all special characters in the string except for punctuation and maintain a list of punctuation characters... Tell me what you prefer

File: assets/scripts/whichx.js (hunk 3 above)
Comment (Sharcoux) [reply assets/scripts/whichx.js:175]: This also document the project for potential contributors, and can detect mistakes through type checking in PRs. If size is a concern, I advise using a minifier and a build process to the project.

File: assets/scripts/whichx.js (hunk 4 above)
Comment (Sharcoux) [reply assets/scripts/whichx.js:204]: See above comment

File: assets/scripts/whichx.js (hunk 5 above)
Comment (Sharcoux) [reply assets/scripts/whichx.js:167]: See above comment

**Comment by Sharcoux** [comment]
About tests, I updated the tests to match the new error messages, so this should be ok.
I'll add a test for the normalization.

File: assets/scripts/whichx.js (hunk 2 above)
Comment (rudikershaw) [reply assets/scripts/whichx.js]: One of the original goals of this project was to leave it without any dependencies (at least not in the deliverable). Additionally, unorm looks like it needs a shim to work pre-ECMA 5. I'm not willing to start using dependencies in the deliverable, how-ever we could simply check if the function is defined before calling it.

```
if (typeof(description.normalize) === typeof(Function)) {
     description = description.normalize("NFD");
}
```

Something like the above, or similar.

File: assets/scripts/whichx.js (hunk 3 above)
Comment (rudikershaw) [reply assets/scripts/whichx.js:175]: I intend to add a minifier some time soon. I will need to look into the specifics, but I had assumed that stripping out the JSDoc in the deliverable will prevent them showing in the IDE. I can look into whether I can publish a sourceMap to NPM, if that is the case.

**Comment by Sharcoux** [comment]
Fix project configuration

 - Add test for diacritic
 - Add generation of type file from JSDoc 
 - Add eslint-plugin-mocha to handle test files
 - Apply eslint on test files
 - Add minify to the dev dependencies for minification and reducing the dependency footprint
 - Complete build process with minify and tsc steps
 - Move code from assets/script to src to match more conventional patterns

**Review by rudikershaw — APPROVED** [review]

**Comment by rudikershaw** [comment]
Thanks again. Changes look good, merging. I will release a 3.0.0 for these changes as the normalization will be backward incompatible with old models.

**Comment by rudikershaw** [comment]
FYI @Sharcoux, a `3.0.1` version has been released that includes these changes.
